src/txn/txn_timestamp.c

```diff
@@ -240,36 +278,38 @@ __txn_global_query_timestamp(
 
 	WT_STAT_CONN_INCR(session, txn_query_ts);
 	WT_RET(__wt_config_gets(session, cfg, "get", &cval));
-	if (WT_STRING_MATCH("all_committed", cval.str, cval.len)) {
-		if (!txn_global->has_commit_timestamp)
+	if (WT_STRING_MATCH("all_committed", cval.str, cval.len) ||
+	    WT_STRING_MATCH("all_durable", cval.str, cval.len)) {
+		if (!txn_global->has_durable_timestamp)
 			return (WT_NOTFOUND);
-		ts = txn_global->commit_timestamp;
+		ts = txn_global->durable_timestamp;
 		WT_ASSERT(session, ts != 0);
 
-		/* Skip the lock if there are no running transactions. */
-		if (TAILQ_EMPTY(&txn_global->commit_timestamph))
+		/*
+		 * Skip straight to the commit queue if no running transactions
+		 * have an explicit durable timestamp.
+		 */
+		if (TAILQ_EMPTY(&txn_global->durable_timestamph))
 			goto done;
-
-		/* Compare with the oldest running transaction. */
-		__wt_readlock(session, &txn_global->commit_timestamp_rwlock);
-		TAILQ_FOREACH(txn, &txn_global->commit_timestamph,
-		    commit_timestampq) {
-			if (txn->clear_commit_q)
+		/*
+		 * Compare with the least recently durable transaction.
+		 */
+		__wt_readlock(session, &txn_global->durable_timestamp_rwlock);
+		TAILQ_FOREACH(txn, &txn_global->durable_timestamph,
+		    durable_timestampq) {
+			if (txn->clear_durable_q)
 				continue;
 
-			tmpts = txn->first_commit_timestamp;
-			WT_ASSERT(session, tmpts != 0);
```

Comment:
This seems like a worthwhile check - we are returning an unsigned value and decrementing below 0 will mean that we return something wrapped to max instead. I know it was only being done in diagnostic mode before - we should probably make this more robust for release builds as well.

Reply:
This assertion still exists but it has been moved within the `__txn_get_published_timestamp` function.

As for making it more robust in release mode, what would you expect WiredTiger to do if it sees something like this?

Reply:
I think it'd be better to return 0 rather than UINT_MAX.

Reply:
As discussed, `all_durable` won't return UINT_MAX here. It'll return `txn_global->durable_timestamp`.

It's hard to make this robust since we call this function when inserting into the queue also and there's no reasonable way to recover from that. I think the only possibilities are to leave it as an assertion or to invoke a panic here.